jpype/_core.py

```diff
@@ -327,17 +331,25 @@ def shutdownJVM():
     restart the JVM after being terminated.
     """
     import threading
+    import jpype.config
     if threading.current_thread() is not threading.main_thread():
         raise RuntimeError("Shutdown must be called from main thread")
-    _jpype.shutdown()
+    if _jpype.isStarted():
+        _jpype.JPypeContext.freeResources = jpype.config.free_resources
+    _jpype.shutdown(jpype.config.destroy_jvm, jpype.config.free_jvm)
 
 
 # In order to shutdown cleanly we need the reference queue stopped
 # otherwise, we can experience a crash if a Java thread is waiting
 # for the GIL.
 def _JTerminate():
     try:
-        _jpype.shutdown()
+        import jpype.config
+        # We are exiting anyway so no need to free resources
+        if _jpype.isStarted():
+            _jpype.JPypeContext.freeResources = False
+        if jpype.config.onexit:
```

Comment:
would it be possible to cover this in the tests?

Reply:
We can hit each path but I am unsure what to test for as these options are mostly about how much memory to free.  Agressively clearing memory just to call exit is an invitation to trigger race conditions.  But what behavior change should should be tested for here?